drivers/usb/device/usb_dc_smartbond.c

```diff
@@ -0,0 +1,1398 @@
+#include <zephyr/usb/usb_device.h>
+
+#include <DA1469xAB.h>
+#include <soc.h>
+#include <da1469x_clock.h>
+
+#include <zephyr/logging/log.h>
+#include <zephyr/drivers/gpio.h>
+
+LOG_MODULE_REGISTER(usb_dc_smartbond, CONFIG_USB_DRIVER_LOG_LEVEL);
+
+/* USB device controller access from devicetree */
+#define DT_DRV_COMPAT renesas_smartbond_usbd
+
+#define USB_IRQ                        DT_INST_IRQ_BY_IDX(0, 0, irq)
+#define USB_IRQ_PRI                    DT_INST_IRQ_BY_IDX(0, 0, priority)
+#define VBUS_IRQ                       DT_INST_IRQ_BY_IDX(0, 1, irq)
+#define VBUS_IRQ_PRI                   DT_INST_IRQ_BY_IDX(0, 1, priority)
+#define DMA_CHANNEL_RX                 DT_INST_PROP(0, dma_chan_rx)
+#define DMA_CHANNEL_TX                 DT_INST_PROP(0, dma_chan_tx)
+/* Minimal transfer size needed to use DMA. For short transfers
+ * it may be simpler to just fill hardware FIFO with data instead
+ * of programming DMA registers.
+ */
+#define DMA_MIN_TRANSFER_SIZE          DT_INST_PROP(0, dma_min_transfer_size)
+#define FIFO_READ_THRESHOLD            DT_INST_PROP(0, fifo_read_threshold)
+
+/* Size of hardware RX and TX FIFO. */
+#define EP0_FIFO_SIZE                  8
+#define EP_FIFO_SIZE                   64
+
+#define EP0_OUT_BUF_SIZE               EP0_FIFO_SIZE
+#define EP1_OUT_BUF_SIZE               DT_INST_PROP_BY_IDX(0, ep_out_buf_size, 1)
+#define EP2_OUT_BUF_SIZE               DT_INST_PROP_BY_IDX(0, ep_out_buf_size, 2)
+#define EP3_OUT_BUF_SIZE               DT_INST_PROP_BY_IDX(0, ep_out_buf_size, 3)
+
+#define EP0_IDX                        0
+#define EP0_IN                         USB_CONTROL_EP_IN
+#define EP0_OUT                        USB_CONTROL_EP_OUT
+#define EP_MAX                         4
+
+/* EP OUT buffers  */
+static uint8_t ep0_out_buf[EP0_OUT_BUF_SIZE];
+static uint8_t ep1_out_buf[EP1_OUT_BUF_SIZE];
+static uint8_t ep2_out_buf[EP2_OUT_BUF_SIZE];
+static uint8_t ep3_out_buf[EP3_OUT_BUF_SIZE];
+
+static uint8_t *const ep_out_bufs[4] = {
+	ep0_out_buf, ep1_out_buf, ep2_out_buf, ep3_out_buf,
+};
+
+static const uint16_t ep_out_buf_size[4] = {
+	EP0_OUT_BUF_SIZE, EP1_OUT_BUF_SIZE, EP2_OUT_BUF_SIZE, EP3_OUT_BUF_SIZE,
+};
+
+/* Node functional states */
+#define NFSR_NODE_RESET         0
+#define NFSR_NODE_RESUME        1
+#define NFSR_NODE_OPERATIONAL   2
+#define NFSR_NODE_SUSPEND       3
+/*
+ * Those two following states are added to allow going out of sleep mode
+ * using frame interrupt.  On remove wakeup RESUME state must be kept for
+ * at least 1ms. It is accomplished by using FRAME interrupt that goes
+ * through those two fake states before entering OPERATIONAL state.
+ */
+#define NFSR_NODE_WAKING        (0x10 | (NFSR_NODE_RESUME))
+#define NFSR_NODE_WAKING2       (0x20 | (NFSR_NODE_RESUME))
+
+struct smartbond_ep_reg_set {
+	__IOM uint32_t epc_in;
+	__IOM uint32_t txd;
+	__IOM uint32_t txs;
+	__IOM uint32_t txc;
+	__IOM uint32_t epc_out;
+	__IOM uint32_t rxd;
+	__IOM uint32_t rxs;
+	__IOM uint32_t rxc;
```

Comment:
NACK for __IOM/__IM macros, AFAIK Keil specific. Others use `volatile uint32_t`.

Reply:
The latest push is updated to use volatile instead of __IOM